Add events view

dorner opened the pull request:
Fixes #4079.

This adds a simple events view for org admins:
<img width="1332" alt="image" src="https://github.com/rubyforgood/human-essentials/assets/1986893/cec48f87-2b9f-424b-9a40-cb955675a980">

Missing some features:
* CSV export
* Filter by storage location and item

We can do that as a fast follow.

cielf commented:
@dorner I was going to ask for this!   Obviously needing the tests to pass, but if we can I'd like to see something like a date on them .   Is it possible for us to say whether it is new or edit?

Also - we need to make the column headers more relatable -- because to them,  "Event" is likely to sound like "the thing I did" -- which to them will be more like what the "resource" is (I'm assuming resource is the itemizable).      What would you think of something like "internal id"  and "Refers to".

cielf commented:
Tried it on a fresh setup on local.  Did not turn on the events.  1/  Was able to get at it -- I had assumed this would be behind the events flag 2/  Got the following:

<img width="765" alt="Screenshot 2024-02-09 at 6 09 30 PM" src="https://github.com/rubyforgood/human-essentials/assets/10157589/3097806e-aa42-41c0-bdcf-e508ca7023db">

I don't know if that's a problem with the seed, or what,  but I'll need that to work to be able to kick the tires enough to give a better opinion.

Thanks.

cielf reviewed (changes requested):
Hey @dorner   Thank you for getting this far.  I'm really hoping we can include this in when we go to early adopters.   I've got an idea or two to make it more approachable for our key users.

 Alas, it feels like maybe this one was a bit of a rush to get in before your week-end deadline  -- it's not quite at your usual standard for getting the i's dotted and the t's crossed.  I'm afraid there are some problems with the tests, with the lint, and with being able to run it on a fresh setup on my local.

dorner commented:
It was definitely rushed, you are 100% correct. Didn't even have time to see if the tests and lint finished before logging off. No problem at all waiting for fixes.

dorner commented:
Btw I didn't put it behind the flag because the events are still _happening_, they're just not being _used_. But if you think it'd be confusing, I can put it behind the flag.

cielf commented:
I think we'll need to rethink the terminology for in front of the users.   A new user is not going to have any idea of what "Event History" means.   Maybe just "History"?       I'd be ok with putting it in front of the early adopters and explaining what it is to each of them as is,   so if you could put it in behind the flag that'd be great.

cielf commented:
At this point, I see this as mainly a helping-with-diagnoses tool,  but it may end up being something someone ends up relying on.

dorner commented:
@cielf I've made the suggested changes. The crash was due to SnapshotEvents having a different format than the rest (not sure why I didn't come across it before).

Should still have a fast-follow where I add more tests and filters.

dorner commented:
I didn't add the event flag because that's not merged yet :) I can put it in as soon as it is.

cielf commented:
Ok...just dipping my toe in the water of kicking the tires.   Here's what I did, what I expected (given the expected use as a troubleshooting tool) and what I see:

I started with a fresh seeded setup,  and then I went and edited a purchase, to add an item to it (since the seed has none).
I expected that I would see something like a note that i edited a purchase, with the new values, at the time that I made the changes.
Instead, I see it at what i assume is the "purchase date", since it is bang-on midnight, which is *not* when the thing happened.    That may be a seed thing, though, because it is the correct date when I add a new one.

cielf commented:
As for the event flag - I think we should have it before it goes live, but it's actually a bit convenient to have it not on at this point.

dorner commented:
Good point - the history date doesn't differentiate between creation and update. I'll put something in there for that.

dorner commented:
Hmm... actually that's a lot harder than it sounds. I think leaving it as is is probably better (i.e. a "Distribution" event could mean creation or update). I'll check the case you mentioned.

cielf commented:
If it doesn't give the whole story, we might want to keep it to ourselves, though.   In my mind its purpose was for walking through what happened step-by-step.  Almost a log.      Let's talk about it Sunday.

dorner commented:
I've added type and storage location filters: 
<img width="1119" alt="image" src="https://github.com/rubyforgood/human-essentials/assets/1986893/64080a53-4397-4b51-af1c-ba32aed82494">

Unfortunately I'm reeeeally behind on tests (i.e. I don't have any for any of this). Hoping to get some of them in on Monday.

dorner commented:
Oh I also added the event flag check.

dorner commented:
@cielf added the item and eventable filters, and added tests.

dorner commented:
Snapshots aren't per-storage-location, they're global. So by design the filter includes all snapshots.

cielf commented:
Another thing I'm wondering is whether we should make the initial date range smaller.   

This is coloured by my picture of it as a troubleshooting tool -- most of the time I would hope we'll just be looking at the last month or two - or at least we would start by looking at the last short while.   

What do you think?

dorner commented:
The default view is unfortunately hardcoded into the helpers. We really should find a way to be able to break that out. :(

cielf commented:
> Snapshots aren't per-storage-location, they're global. So by design the filter includes all snapshots.

But they say that they belong to a storage location in the view....

dorner commented:
They have multiple storage locations. If you scroll down you'll see both. There isn't a great way to split that up in the view.

cielf commented:
Hmmm.   How often are we doing snapshots?  Just to start the thing off, really?

dorner commented:
For now. We probably will introduce more in the future for performance purposes.

cielf commented:
Ok.   I think that might end up being confusing to the users, but I'm willing to early adopt this as it is and kick that particular can down the road until we see if people end up using this for anything but troubleshooting.  

I'd like @awwaiid to take a look at this from a technical pov.  (Though I'm pretty sure it will pass with flying colours once you clear the failing tests/lint.)

dorner commented:
@cielf I added a separate background color for snapshot events which should at least make it more obvious that the two rows are grouped together. 
<img width="1054" alt="image" src="https://github.com/rubyforgood/human-essentials/assets/1986893/f23fd832-dbe3-497c-a6d9-b745afabe1f1">

Lint and tests should be fixed.

cielf commented:
(nods)  I can foresee a future, if this is useful to the banks for something other than this troubleshooting, where we hide the snapshots by default, and have an "include system snapshots" checkbox,  but that's good for now.

cielf commented:
@awwaiid  Can you do a quick pass on this from a technical pov.  LGTM from a functional pov.  Thank you.

cielf commented:
@dorner Ok.  This is late-breaking,  but we've got the user information on the event, yes?  That could actually be handy when we are talking through troubleshooting with the banks.  *But* I'm totally willing to have it be a fast follow-up in the interest of getting this in.

dorner commented:
I've replaced Internal ID with user name. The ID isn't helpful to anyone but us. If we really need it, I added it as an HTML tooltip on the "event type" column. 
<img width="1208" alt="image" src="https://github.com/rubyforgood/human-essentials/assets/1986893/d3d0627c-7da1-4386-8ba8-af35a05c28f6">

cielf commented:
(nods) That sounds reasonable.  Thank you.

awwaiid commented on config/initializers/event_types.rb:
```diff
@@ -0,0 +1,5 @@
+# load all event types so we can use it in e.g. drop-downs.
+
+Rails.application.reloader.to_prepare do
+  Dir["#{Rails.root}/app/events/*_event.rb"].each { |f| require f }
```
Can/should this be done with `config.eager_load_paths`?

dorner replied:
Tried it and it doesn't seem to work, at least on development.

awwaiid commented on spec/services/distribution_itemized_breakdown_service_spec.rb:
```diff
@@ -33,13 +33,11 @@
-      end
+      expected_output_csv = <<~CSV
+        Item,Total Distribution,Total On Hand
+        Item 1,500,100
+        Item 2,200,200
```
These are named `A Diapers` and `B Diapers`. Spec might pass with:
```suggestion
        A Diapers,500,100
        B Diapers,200,200
```

awwaiid reviewed (changes requested):
Good overall, but actual spec fails and I'd like to know about the eager-load.

awwaiid commented:
I fixed the one easy spec error, but I'm not sure on the other one -- looks like events without users.

dorner commented:
Should be fixed now I hope!
